action/protocol/staking/handlers_test.go

```diff
@@ -166,6 +167,1624 @@ func TestProtocol_HandleCreateStake(t *testing.T) {
+			require.NoError(err)
+			actCost, err := act.Cost()
+			require.NoError(err)
+			cuCost, err := cu.Cost()
+			require.NoError(err)
+			total := big.NewInt(0)
+			require.Equal(unit.ConvertIotxToRau(test.initBalance), total.Add(total, caller.Balance).Add(total, actCost).Add(total, cuCost).Add(total, p.config.RegistrationConsts.Fee))
+			require.Equal(test.nonce, caller.Nonce)
+		}
+	}
+}
+
+func TestProtocol_HandleUnstake(t *testing.T) {
+	require := require.New(t)
+	ctrl := gomock.NewController(t)
+	defer ctrl.Finish()
+
+	sm, p, candidate, candidate2 := initAll(t, ctrl)
+	initCreateStake(t, sm, identityset.Address(2), 100, big.NewInt(unit.Qev), 10000, 1, 1, time.Now(), 10000, p, candidate2, "10000000000000000000", false)
+
+	callerAddr := identityset.Address(1)
+	tests := []struct {
+		// creat stake fields
+		caller      address.Address
+		amount      string
+		initBalance int64
+		selfstaking bool
+		// action fields
+		index    uint64
+		gasPrice *big.Int
+		gasLimit uint64
+		nonce    uint64
+		// block context
+		blkHeight    uint64
+		blkTimestamp time.Time
+		blkGasLimit  uint64
+		// clear flag for inMemCandidates
+		clear bool
+		// need new p
+		newProtocol bool
+		err         error
+		// expected result
+		status iotextypes.ReceiptStatus
+	}{
+		// fetchCaller ErrNotEnoughBalance
+		{
+			callerAddr,
+			"9990000000000000000",
+			10,
+			false,
+			0,
+			big.NewInt(unit.Qev),
+			10000,
+			1,
+			1,
+			time.Now(),
+			10000,
+			false,
+			true,
+			nil,
+			iotextypes.ReceiptStatus_ErrNotEnoughBalance,
+		},
+		// fetchBucket, bucket.Owner is not equal to actionCtx.Caller
+		{
+			identityset.Address(12),
+			"10000000000000000000",
+			100,
+			false,
+			0,
+			big.NewInt(unit.Qev),
+			10000,
+			1,
+			1,
+			time.Now(),
+			10000,
+			false,
+			false,
+			nil,
+			iotextypes.ReceiptStatus_ErrUnauthorizedOperator,
+		},
+		// fetchBucket,ReceiptStatus_ErrInvalidBucketType cannot happen,because allowSelfStaking is true
+		// fetchBucket and updateBucket call getbucket, ReceiptStatus_ErrInvalidBucketIndex
+		{
+			identityset.Address(33),
+			"10000000000000000000",
+			100,
+			false,
+			1,
+			big.NewInt(unit.Qev),
+			10000,
+			1,
+			1,
+			time.Now(),
+			10000,
+			false,
+			true,
+			nil,
+			iotextypes.ReceiptStatus_ErrInvalidBucketIndex,
+		},
+		// inMemCandidates.GetByOwner,ErrInvalidOwner
+		{
+			callerAddr,
+			"10000000000000000000",
+			100,
+			false,
+			0,
+			big.NewInt(unit.Qev),
+			10000,
+			1,
+			1,
+			time.Now(),
+			10000,
+			true,
+			true,
+			ErrInvalidOwner,
+			iotextypes.ReceiptStatus_Success,
+		},
+		// Upsert error cannot happen,because collision cannot happen
+		// success
+		{
+			callerAddr,
+			"10000000000000000000",
+			100,
+			false,
+			0,
+			big.NewInt(unit.Qev),
+			10000,
+			1,
+			1,
+			time.Now(),
+			10000,
+			false,
+			true,
+			nil,
+			iotextypes.ReceiptStatus_Success,
+		},
+	}
+
+	for _, test := range tests {
+		if test.newProtocol {
```

Comment:
why have this? to clear the sm?

Reply:
for test inMemCandidates.GetByOwner,this error cannot produce if dont delete candidate in "inMemCandidates" cache

Reply:
right,this flag is for "fetchBucket, bucket.Owner is not equal to actionCtx.Caller" error.
That can only produce in the same sm